Allow Dataflow flex template jobs to be updated

andremarianiello opened the pull request:
I really need https://github.com/hashicorp/terraform-provider-google/issues/8408, so this is my attempt at adding support for it. I would be happy to add an update test as per the contribution guidelines, but the existing test jobs create jobs that fail to run due to lack of a real template image, and updates will not work without a running job. Should I fix this somehow?

andremarianiello commented:
Is there anything I need to do to help usher this through the approval process?

slevenick commented:
Yeah, if we need to be able to set up correctly running jobs to update them it would definitely help if you could figure out what needs to change for the test jobs to correctly start.

It probably just requires a valid image to be used?

andremarianiello commented:
Yeah, I believe I can just update the tests to use a valid image. The best way to do it would to simply make a publicly available image that anyone who runs the tests can use. Is this an acceptable approach?

slevenick commented:
> Yeah, I believe I can just update the tests to use a valid image. The best way to do it would to simply make a publicly available image that anyone who runs the tests can use. Is this an acceptable approach?

It would be better to use an already public image that is hosted somewhere. Does one of these exist, maybe in a GCP guide on setting up these templates?

andremarianiello commented:
Google does provide quite a few premade flex templates in gs://dataflow-templates/latest/flex/, and I think I can run some of them without having to create any other resources, so that is the approach I will try. Thanks for the idea!

andremarianiello commented:
@slevenick I have updated the existing flex template tests to use the Streaming Data Generator flex template located in the dataflow-templates gcs bucket. When run they continuously produce errors when trying to access non-existent cloud resources, but the job remains in the running state, which is all that is required. I also added a test called TestAccDataflowFlexTemplateJob_streamUpdate which tests the new job update function.

andremarianiello commented:
The CI failed on a file I didn't touch. Do I have to worry about that?

andremarianiello commented:
I triggered the CI with no changes and it passed. Seems like its just a flaky check to me. Please let me know if you would like any changes.

slevenick commented:
I'm working to upstream this to our code generation tool so it gets preserved between runs. It seems like these flex template jobs are only allowed to be launched, not updated. This implementation seems to launch a new job with the same name and overwrite the existing job in state. This seems like it would abandon the existing job in favor of the "updated" job which is actually a new job. Is this interpretation correct?

I'm wondering why you need update support, rather than the existing delete & recreate functionality, as by my read this PR does essentially the same thing

slevenick commented:
Disregard the above comment, just saw the `Update:               true,` flag in the update request. Odd API functionality

andremarianiello commented:
Yup, it is definitely very weird. If you start a job with the update flag and the same name, Dataflow will migrate persisted data from the old job to the new one.

slevenick commented on google-beta/resource_dataflow_job.go:
```diff
@@ -580,7 +580,7 @@ func resourceDataflowJobIterateMapHasChange(mapKey string, d *schema.ResourceDat
 }
 
-func resourceDataflowJobIsVirtualUpdate(d *schema.ResourceData) bool {
+func resourceDataflowJobIsVirtualUpdate(d *schema.ResourceData, resourceSchema map[string]*schema.Schema) bool {
```
How is this supposed to work when `resourceSchema` is never used? I'm unclear on why this method is being changed

andremarianiello replied:
Good catch! I forgot to remove the `resourceSchema := resourceDataflowJob().Schema` line. This change was supposed to take the resource schema to use as a parameter so that it can be used for both job and flex job resource schemas. I will push a fix for this issue.

andremarianiello replied:
Alright, I tested and pushed the change. I simply deleted the line that was shadowing the `resourceSchema` parameter.

slevenick commented on google-beta/resource_dataflow_flex_template_job.go:
```diff
@@ -177,10 +174,96 @@ func resourceDataflowFlexTemplateJobRead(d *schema.ResourceData, meta interface{
 func resourceDataflowFlexTemplateJobUpdate(d *schema.ResourceData, meta interface{}) error {
-	return nil
+	// Don't send an update request if only virtual fields have changes
+	if resourceDataflowJobIsVirtualUpdate(d, resourceDataflowFlexTemplateJob().Schema) {
```
I don't think this resource has virtual fields, so this can probably be removed

andremarianiello replied:
Just like with dataflow_job, dataflow_flex_template_job has a virtual "on_delete" field. It is not used when creating or updating a job. It is only used when the job resource should be deleted, and governs whether to drain or cancel the job (the two ways of stopping a running job).

slevenick commented on google-beta/resource_dataflow_flex_template_job.go:
```diff
@@ -177,10 +174,96 @@ func resourceDataflowFlexTemplateJobRead(d *schema.ResourceData, meta interface{
+	}
+
+	job := response.Job
+	d.SetId(job.Id)
```
Does the `id` of the job change when updating?

andremarianiello replied:
Yup, it does. It is similar to the logic in dataflow_job resource updates. If you launch a flex template job it will get a unique ID regardless of whether or not it is an update.

slevenick commented on google-beta/resource_dataflow_flex_template_job_test.go:
```diff
@@ -123,136 +151,87 @@ func testAccDataflowFlexTemplateJobGetGeneratedInstance(t *testing.T, s *terrafo
   parameters = {
-    inputSubscription = "my-subscription"
-    outputTable  = "my-project:my-dataset.my-table"
+    schemaLocation = "gs://mybucket/schema.json"
```
Does this need to reference real objects and topics? This seems like it wouldn't successfully run

andremarianiello replied:
Luckily, no, it doesn't need to reference real cloud resources. All the is required to update a job is that the existing job be in a RUNNING state. Using fake resource names for sources/sinks in the job causes the source and sink stages to emit errors while running, but the pipeline as a whole is still running.

andremarianiello commented:
@slevenick Would you like me to resolve conversations that I believe I have fully addressed, or would you prefer that I let you do that?

slevenick commented:
Looks fine to me, let me get a review on the upstream PR and we should be good to go

slevenick reviewed (approved):
Thanks! And sorry I missed final review & merge for so long

I've upstreamed this into our code generator so that will take care of keeping this synced between the beta & GA providers

andremarianiello commented:
> Thanks! And sorry I missed final review & merge for so long
> 
> I've upstreamed this into our code generator so that will take care of keeping this synced between the beta & GA providers

No worries, thanks for taking the time! Thanks also for upstreaming the changes. I'm glad I didn't have to figure out how to use the upstream repo.
